### PR TITLE
AE-1313 Incorporate Autotasks into Compound Repo

### PR DESCRIPTION
**Pull request by darian-ae**
This PR incorporates the autotask code for sending discord alerts when Openzepplin's Defender Forta Sentinels pick up on an alert from out Forta Bots. Not all bots have autotasks written for them (the newer bots), but autotasks can be found for:

- asset monitor
- cToken transaction monitor
- COMP distribution monitor
- governance monitor
- large borrows governance monitor
- oracle price monitor

The code for the autotask can be found under the `src` folder for each Bot, as well as the associated tests. Note that the majority of these bots have not picked up alerts yet (as they were deployed beginning of April). This made it difficult to test using Forta's API with their real data, so a lot of the data in the `autotask-spec.js` files have been mocked locally.

### CODE REVIEW

File: asset-monitor/src/autotask.js [hunk 1]
```diff
@@ -0,0 +1,186 @@
+//  CTOKEN UNDERLYING ASSET MONITOR AGENT ALERT - note there's never been an alert to this agent yet
+// agent id - 0x3f02bee8b17edc945c5c1438015aede79225ac69c46e9cd6cff679bb71f35576
+
+const axios = require('axios');
+const needle = require('needle');
```
Comment (Nomad-AE) [inline asset-monitor/src/autotask.js:5]: Any reason that you are using `axios.post` and needle.post`? They both can do the same thing

File: ctoken-monitor/src/autotask.js [hunk 2]
```diff
@@ -0,0 +1,330 @@
...
+    response = await post(url, method, headers, data);
+  } catch (error) {
+    // is this a "too many requests" error (HTTP status 429)
+    if (error.response && error.response.status === 429) {
+      // the request was made and a response was received
+      // try again after waiting 5 seconds
+      // eslint-disable-next-line no-promise-executor-return
+      const promise = new Promise((resolve) => setTimeout(resolve, 5000));
+      await promise;
+      response = await post(url, method, headers, data);
```
Comment (Nomad-AE) [inline ctoken-monitor/src/autotask.js:164]:
```suggestion
      try {
        response = await post(url, method, headers, data);
      } catch (error2) {
        // Handle error
      }
```
What if this errors out?

Comment (Nomad-AE) [reply ctoken-monitor/src/autotask.js:164]: Actually, your way is good. Ref: https://eslint.org/docs/rules/no-useless-catch

File: ctoken-monitor/README.md [hunk 3]
```diff
@@ -31,3 +31,37 @@ Liquidate (DAI) - 0x4316c6cb73296b6b5b603a80a267b203d394e049216d71d8e3ff4822e8d7
 Mint (USDC) - 0xfd200d6d3136ce4f4b91270f8ed7579bf9932e1ce2388b324ee4a281f58047b1
 Redeem (ETH) - 0x4b11cee2963cbb2c01cf2c9858a684117ef24ce62a4ffac5738557e14f0276b6
 Repay (USDC) - 0x3c59e9ba77e3864f2efdc49df1c9aba83fb202c59ffbbe116c133dbeb327b010
+
+## Autotask
+
+This autotask sends alerts from Openzepplin's Defender Forta Sentinel's to Compound's Disord channel.
```
Comment (Nomad-AE) [inline ctoken-monitor/README.md:37]:
```suggestion
This autotask sends alerts from Openzepplin's Defender Forta Sentinel's to Compound's Discord channel. 
```

File: ctoken-monitor/README.md [hunk 4]
```diff
@@ -31,3 +31,37 @@ Liquidate (DAI) - 0x4316c6cb73296b6b5b603a80a267b203d394e049216d71d8e3ff4822e8d7
...
+- Create an `autotask-config.json` file with the following format:
+```
+{
+  "jsonRpcUrl": "https://your.preferred.json.rpc.endpoint/with/api-key",
+  "agentId": "0xFORTAAGENTIDHERE",
+  "startBlockNumber": <integer_for_starting_block_number>,
+  "endBlockNumber": <integer_for_ending_block_number>
+}
+```
+- Create a .env file that contains the name of your discord webook and the url for it:
```
Comment (Nomad-AE) [inline ctoken-monitor/README.md:66]:
```suggestion
- Create a .env file that contains the name of your discord webhook and the URL for it:
```
Acronyms should be uppercased

File: distribution/README.md [hunk 5]
```diff
@@ -27,4 +27,38 @@ The following transactions can be used to test the operation of this agent.
 0xbc246c878326f2c128462d08a0b74048b1dbee733adde8863f569c949c06422a
 
 To run, use:
-`npm run tx {transactionHash}`
\ No newline at end of file
+`npm run tx {transactionHash}`
+
+## Autotask
+
+This autotask sends alerts from Openzepplin's Defender Forta Sentinel's to Compound's Disord channel.
```
Comment (arbitrary-sparrow) [inline distribution/README.md:34]:
```suggestion
This Autotask sends alerts from Openzeppelin's Defender Forta Sentinels to Compound's Discord channel. 
```

File: distribution/src/autotask.js [hunk 6]
```diff
@@ -0,0 +1,220 @@
+// DISTRIBUTIONS AGENT ALERT
+// agent id - 0xfca83adc900f88f22dafcd91117d0929343cba3f18e4607bcd861ff0bcd706fa
```
Comment (arbitrary-sparrow) [inline distribution/src/autotask.js:2]:
```suggestion
```

File: governance/README.md [hunk 7]
```diff
@@ -88,3 +88,37 @@ VoteCast - 0xd3b303df1c7cd21bc8442f436cda91a1ee00872eab212d74ba5ccd8ea067c760
 
 To run, use:
 `npm run tx {transactionHash}`
+
+## Autotask
+
+This autotask sends alerts from Openzepplin's Defender Forta Sentinel's to Compound's Disord channel.
```
Comment (arbitrary-sparrow) [inline governance/README.md:94]:
```suggestion
This Autotask sends alerts from Openzeppelin's Defender Forta Sentinels to Compound's Discord channel. 
```

File: governance/src/autotask.js [hunk 8]
```diff
@@ -0,0 +1,311 @@
+// eslint-disable-next-line import/no-extraneous-dependencies
+// agentID - 0xc6c224fb57f6bf4c14b88afc1b3e8ec44522363abb740d3d19915837bc57d605
```
Comment (arbitrary-sparrow) [inline governance/src/autotask.js:2]:
```suggestion
```

File: large-borrows-governance/README.md [hunk 9]
```diff
@@ -23,4 +23,38 @@ to pass significant governance thresholds. This can be an early indication of go
 Run unit tests:
 ```
 npm test
-```
\ No newline at end of file
+```
+
+## Autotask
+
+This autotask sends alerts from Openzepplin's Defender Forta Sentinel's to Compound's Disord channel.
```
Comment (arbitrary-sparrow) [inline large-borrows-governance/README.md:30]:
```suggestion
This Autotask sends alerts from Openzeppelin's Defender Forta Sentinels to Compound's Discord channel. 
```

File: large-borrows-governance/src/autotask.js [hunk 10]
```diff
@@ -0,0 +1,195 @@
+// LARGE BORROWS GOVERNANCE AGENT ALERT - note there's never been an alert to this agent yet
+// agent id - 0x704a35d97dfae2caf3b409206f1bc4c8a06671170108d07f9b3c7ea5026916fc
```
Comment (arbitrary-sparrow) [inline large-borrows-governance/src/autotask.js:2]:
```suggestion
```

File: oracle-price-monitor/README.md [hunk 11]
```diff
@@ -28,4 +28,38 @@ npx forta-agent run --tx 0xe9456ccee1b1764dfe80291f3b894a29f0789f20f995de7d88ff1
 Run unit tests:
 ```
 npm test
-```
\ No newline at end of file
+```
+
+## Autotask
+
+This autotask sends alerts from Openzepplin's Defender Forta Sentinel's to Compound's Disord channel.
```
Comment (arbitrary-sparrow) [inline oracle-price-monitor/README.md:35]:
```suggestion
This Autotask sends alerts from Openzeppelin's Defender Forta Sentinels to Compound's Discord channel. 
```

File: oracle-price-monitor/src/autotask.js [hunk 12]
```diff
@@ -0,0 +1,270 @@
+// ORACLE PRICE ALERT - note there's never been an alert to this agent yet
+// agent id - 0x8a6e2ae3b279c561f7ba64e5011313df19e08b1f398d59d94dfbd4148e6cafce
```
Comment (arbitrary-sparrow) [inline oracle-price-monitor/src/autotask.js:2]:
```suggestion
```

**Review by arbitrary-sparrow — CHANGES_REQUESTED** [review]
Spelling nits and other small changes

File: ctoken-monitor/src/autotask.spec.js [hunk 13]
```diff
@@ -0,0 +1,175 @@
...
+  });
+
+  return autotaskEvents;
+}
+
+it('Runs autotask against blocks in configuration file', async () => {
+  // get the development configuration values
+  const { agentId, startBlockNumber, endBlockNumber } = config;
+
+
```
Comment (Nomad-AE) [inline ctoken-monitor/src/autotask.spec.js:166]:
```suggestion

```

File: ctoken-monitor/src/autotask.spec.js [hunk 14]
```diff
@@ -0,0 +1,175 @@
...
+  // grab Forta Agent alerts from the Forta Public API and create autotaskEvents
+  const autotaskEvents = await createFortaSentinelEvents(agentId, startBlockNumber, endBlockNumber);
+  console.log("autotask event here", autotaskEvents)
```
Comment (Nomad-AE) [inline ctoken-monitor/src/autotask.spec.js:169]:
```suggestion
  console.log('autotask event here', autotaskEvents)
```

File: ctoken-monitor/src/autotask.spec.js [hunk 15]
```diff
@@ -0,0 +1,175 @@
...
+        },
+      },
+    };
+    return autotaskEvent;
+  });
+
+  return autotaskEvents;
+}
+
+it('Runs autotask against blocks in configuration file', async () => {
```
Comment (Nomad-AE) [inline ctoken-monitor/src/autotask.spec.js:162]: Task runs and posts on discord but the test reports `Failure`.  This is the error message `Jest did not exit one second after the test run has completed.

This usually means that there are asynchronous operations that weren't stopped in your tests. Consider running Jest with `--detectOpenHandles` to troubleshoot this issue.`

File: distribution/README.md [hunk 16]
```diff
@@ -27,4 +27,38 @@ The following transactions can be used to test the operation of this agent.
...
+- Create an `autotask-config.json` file with the following format:
+```
+{
+  "jsonRpcUrl": "https://your.preferred.json.rpc.endpoint/with/api-key",
+  "agentId": "0xFORTAAGENTIDHERE",
+  "startBlockNumber": <integer_for_starting_block_number>,
+  "endBlockNumber": <integer_for_ending_block_number>
+}
+```
+- Create a .env file that contains the name of your discord webook and the url for it:
```
Comment (Nomad-AE) [inline distribution/README.md:63]:
```suggestion
- Create a .env file that contains the name of your discord webhook and the URL for it:
```

File: governance/README.md [hunk 17]
```diff
@@ -88,3 +88,37 @@ VoteCast - 0xd3b303df1c7cd21bc8442f436cda91a1ee00872eab212d74ba5ccd8ea067c760
...
+- Create an `autotask-config.json` file with the following format:
+```
+{
+  "jsonRpcUrl": "https://your.preferred.json.rpc.endpoint/with/api-key",
+  "agentId": "0xFORTAAGENTIDHERE",
+  "startBlockNumber": <integer_for_starting_block_number>,
+  "endBlockNumber": <integer_for_ending_block_number>
+}
+```
+- Create a .env file that contains the name of your discord webook and the url for it:
```
Comment (Nomad-AE) [inline governance/README.md:123]:
```suggestion
- Create a .env file that contains the name of your discord webhook and the URL for it:
```

File: large-borrows-governance/.gitignore [hunk 18]
```diff
@@ -3,3 +3,4 @@
 forta.config.json
 node_modules
 dist
+autotask-config.json
```
Comment (Nomad-AE) [inline large-borrows-governance/.gitignore:6]:
```suggestion
autotask-config.json

```

File: large-borrows-governance/README.md [hunk 19]
```diff
@@ -23,4 +23,38 @@ to pass significant governance thresholds. This can be an early indication of go
...
+- Create an `autotask-config.json` file with the following format:
+```
+{
+  "jsonRpcUrl": "https://your.preferred.json.rpc.endpoint/with/api-key",
+  "agentId": "0xFORTAAGENTIDHERE",
+  "startBlockNumber": <integer_for_starting_block_number>,
+  "endBlockNumber": <integer_for_ending_block_number>
+}
+```
+- Create a .env file that contains the name of your discord webook and the url for it:
```
Comment (Nomad-AE) [inline large-borrows-governance/README.md:59]:
```suggestion
- Create a .env file that contains the name of your discord webhook and the URL for it:
```

File: oracle-price-monitor/.gitignore [hunk 20]
```diff
@@ -3,3 +3,4 @@
 forta.config.json
 node_modules
 dist
+autotask-config.json
```
Comment (Nomad-AE) [inline oracle-price-monitor/.gitignore:6]:
```suggestion
autotask-config.json

```

File: oracle-price-monitor/README.md [hunk 21]
```diff
@@ -28,4 +28,38 @@ npx forta-agent run --tx 0xe9456ccee1b1764dfe80291f3b894a29f0789f20f995de7d88ff1
...
+- Create an `autotask-config.json` file with the following format:
+```
+{
+  "jsonRpcUrl": "https://your.preferred.json.rpc.endpoint/with/api-key",
+  "agentId": "0xFORTAAGENTIDHERE",
+  "startBlockNumber": <integer_for_starting_block_number>,
+  "endBlockNumber": <integer_for_ending_block_number>
+}
+```
+- Create a .env file that contains the name of your discord webook and the url for it:
```
Comment (Nomad-AE) [inline oracle-price-monitor/README.md:64]:
```suggestion
- Create a .env file that contains the name of your discord webhook and the URL for it:
```

File: ctoken-monitor/src/autotask.spec.js (hunk 15 above)
Comment (Nomad-AE) [reply ctoken-monitor/src/autotask.spec.js:162]: I've tested with extending the timeout and it still fails. Actually running the normal `npm test` returns successful at first until it too times out. I'm not sure if fixing this issue is within your scope.

File: oracle-price-monitor/src/autotask.js [hunk 22]
```diff
@@ -0,0 +1,270 @@
...
+];
+const CTOKEN_ABI = ['function underlying() view returns (address)'];
+
+const makerTokenAddress = '0x9f8F72aA9304c8B593d555F12eF6589cC3A579A2'.toLowerCase();
+const saiTokenAddress = '0x89d24A6b4CcB1B6fAA2625fE562bDD9a23260359'.toLowerCase();
+const oddTokens = [makerTokenAddress, saiTokenAddress];
+
+const fortaApiEndpoint = 'https://api.forta.network/graphql';
+
+async function getDecimalsAndSymbol(cTokenAddress, provider) {
```
Comment (Nomad-AE) [inline oracle-price-monitor/src/autotask.js:27]:
```suggestion
async function getDecimalsAndSymbol(cTokenAddress, provider) {
  // Special cETH / ETH  case
  if (cTokenAddress.toLowerCase() === '0x4Ddc2D193948926D02f9B1fE9e1daa0718270ED5'.toLowerCase()) {
    const decimals = 18;
    const symbol = 'ETH';
    return { decimals, symbol };
  }
```

File: oracle-price-monitor/src/autotask.js [hunk 23]
```diff
@@ -0,0 +1,270 @@
...
+  const cTokenContract = new ethers.Contract(
+    cTokenAddress,
+    CTOKEN_ABI,
+    provider,
+  );
+  const underlyingTokenAddress = await cTokenContract.underlying();
+
+  let decimals;
+  let symbol;
+  if (oddTokens.indexOf(underlyingTokenAddress.toLowerCase()) !== -1) {
```
Comment (Nomad-AE) [inline oracle-price-monitor/src/autotask.js:37]:
```suggestion
  if (oddTokens.includes(underlyingTokenAddress.toLowerCase())) {
```
nit

File: oracle-price-monitor/src/autotask.js [hunk 24]
```diff
@@ -0,0 +1,270 @@
...
+      provider,
+    );
+    decimals = await underlyingTokenContract.decimals();
+    symbol = await underlyingTokenContract.symbol();
+  }
+  return { decimals, symbol };
+}
+
+function formatAmountString(amount, decimals) {
+  console.log("amount here", amount)
```
Comment (Nomad-AE) [inline oracle-price-monitor/src/autotask.js:64]:
```suggestion
  console.log('amount here', amount)
```
Ref: https://github.com/airbnb/javascript#strings--quotes

File: oracle-price-monitor/src/autotask.js [hunk 25]
```diff
@@ -0,0 +1,270 @@
...
+  const internationalNumberFormat = new Intl.NumberFormat('en-US');
+  return internationalNumberFormat.format(resultString);
+}
+
+async function createDiscordMessage(reporterPrice, cTokenAddress, transactionHash, provider) {
+  const { decimals, symbol } = await getDecimalsAndSymbol(cTokenAddress, provider);
+
+  const amountString = formatAmountString(reporterPrice, decimals);
+
+  console.log("amount string here", amountString)
```
Comment (Nomad-AE) [inline oracle-price-monitor/src/autotask.js:87]:
```suggestion
  console.log('amount string here', amountString);
```
Ref: https://github.com/airbnb/javascript#strings--quotes

File: oracle-price-monitor/src/autotask.js [hunk 26]
```diff
@@ -0,0 +1,270 @@
+// ORACLE PRICE ALERT - note there's never been an alert to this agent yet
```
Comment (Nomad-AE) [inline oracle-price-monitor/src/autotask.js:1]: Lint this code to fix 18 style issues.

File: oracle-price-monitor/src/autotask.js [hunk 27]
```diff
@@ -0,0 +1,270 @@
...
+              number
+              chainId
+            }
+            agent {
+              id
+            }
+          }
+          severity
+		  metadata
+		  description
```
Comment (Nomad-AE) [inline oracle-price-monitor/src/autotask.js:153]:
```suggestion
      metadata
      description
```
Don't mix tabs with spaces.

File: oracle-price-monitor/src/autotask.js [hunk 28]
```diff
@@ -0,0 +1,270 @@
...
+      transactionHash,
+      agent: {
+        id: agentId,
+      },
+    },
+  } = alert;
+
+  // retrieve the metadata from the Forta public API
+  let alerts = await getFortaAlerts(agentId, transactionHash);
+  // alerts = alerts.filter((alertObject) => alertObject.hash === hash);
```
Comment (Nomad-AE) [inline oracle-price-monitor/src/autotask.js:240]:
```suggestion
  const alerts = await getFortaAlerts(agentId, transactionHash);
```
If you don't need this sanity check, then remove it.

File: oracle-price-monitor/src/autotask.js [hunk 29]
```diff
@@ -0,0 +1,270 @@
...
+
+  // ensure that the alert key exists within the body Object
+  const { alert } = body;
+  if (alert === undefined) {
+    return {};
+  }
+
+  // extract the transaction hash and agent ID from the alert Object
+  const {
+    hash,
```
Comment (Nomad-AE) [inline oracle-price-monitor/src/autotask.js:229]:
```suggestion
```
If you don't need this sanity check, then remove it.

File: oracle-price-monitor/src/autotask.js [hunk 30]
```diff
@@ -0,0 +1,270 @@
...
+    response = needle.post(url, {content: message}, {json: true} )
+  } catch (error) {
+    // is this a "too many requests" error (HTTP status 429)
+    if (error.response && error.response.status === 429) {
+      // the request was made and a response was received
+      // try again after waiting 5 seconds
+      // eslint-disable-next-line no-promise-executor-return
+      const promise = new Promise((resolve) => setTimeout(resolve, 5000));
+      await promise;
+      response = needle.post(url, {content: message}, {json: true} )
```
Comment (Nomad-AE) [inline oracle-price-monitor/src/autotask.js:109]: Replace `needle.post` with `axios`. It's unnecessary to have to modules that do the same thing and we use `axios` everywhere else.

File: oracle-price-monitor/src/autotask.spec.js [hunk 31]
```diff
@@ -0,0 +1,155 @@
...
+    return autotaskEvent;
+  });
+
+  return autotaskEvents;
+}
+
+it('Runs autotask against blocks in configuration file', async () => {
+  // get the development configuration values
+  const { agentId, startBlockNumber, endBlockNumber } = autotaskConfig;
+  console.log("agent id", agentId)
```
Comment (Nomad-AE) [inline oracle-price-monitor/src/autotask.spec.js:145]:
```suggestion
  console.log('agent id', agentId);
```

File: oracle-price-monitor/src/autotask.spec.js [hunk 32]
```diff
@@ -0,0 +1,155 @@
...
+
+  // grab Forta Agent alerts from the Forta Public API and create autotaskEvents
+  const autotaskEvents = await createFortaSentinelEvents(agentId, startBlockNumber, endBlockNumber);
+  console.log("hits here")
```
Comment (Nomad-AE) [inline oracle-price-monitor/src/autotask.spec.js:149]:
```suggestion
  console.log('hits here');
```

File: large-borrows-governance/src/autotask.js [hunk 33]
```diff
@@ -0,0 +1,195 @@
+// LARGE BORROWS GOVERNANCE AGENT ALERT - note there's never been an alert to this agent yet
```
Comment (Nomad-AE) [inline large-borrows-governance/src/autotask.js:1]: Lint to remove 13 style issues.

File: large-borrows-governance/src/autotask.js [hunk 34]
```diff
@@ -0,0 +1,195 @@
...
+    response = needle.post(url, {content: message}, {json: true} )
+  } catch (error) {
+    // is this a "too many requests" error (HTTP status 429)
+    if (error.response && error.response.status === 429) {
+      // the request was made and a response was received
+      // try again after waiting 5 seconds
+      // eslint-disable-next-line no-promise-executor-return
+      const promise = new Promise((resolve) => setTimeout(resolve, 5000));
+      await promise;
+      response = needle.post(url, {content: message}, {json: true} )
```
Comment (Nomad-AE) [inline large-borrows-governance/src/autotask.js:38]: Replace `needle.post` with `axios`. It's unnecessary to have to modules that do the same thing and we use `axios` everywhere else.

File: large-borrows-governance/src/autotask.js [hunk 35]
```diff
@@ -0,0 +1,195 @@
...
+              number
+              chainId
+            }
+            agent {
+              id
+            }
+          }
+          severity
+		  metadata
+		  description
```
Comment (Nomad-AE) [inline large-borrows-governance/src/autotask.js:83]:
```suggestion
          metadata
          description
```
Replace tabs with spaces

File: large-borrows-governance/src/autotask.spec.js [hunk 36]
```diff
@@ -0,0 +1,155 @@
+const ethers = require('ethers');
```
Comment (Nomad-AE) [inline large-borrows-governance/src/autotask.spec.js:1]: Lint to remove 5 style issues.

File: large-borrows-governance/src/autotask.spec.js [hunk 37]
```diff
@@ -0,0 +1,155 @@
...
+
+// DEFENDER FORTA SENTINEL AUTOTASK TESTING //
+
+// grab the existing keys before loading new content from the .env file
+const existingKeys = Object.keys(process.env);
+require('dotenv').config();
+
+// now filter out all of the existing keys from what is currently in the process.env Object
+const newKeys = Object.keys(process.env).filter((key) => existingKeys.indexOf(key) === -1);
```
Comment (Nomad-AE) [inline large-borrows-governance/src/autotask.spec.js:10]:
```suggestion
const newKeys = Object.keys(process.env).filter((key) => !existingKeys.includes(key));
```
Nit. Includes does what you want to do.

File: distribution/src/autotask.spec.js [hunk 38]
```diff
@@ -0,0 +1,175 @@
...
+
+const ethers = require('ethers');
+const axios = require('axios');
+
+// grab the existing keys before loading new content from the .env file
+const existingKeys = Object.keys(process.env);
+require('dotenv').config();
+
+// now filter out all of the existing keys from what is currently in the process.env Object
+const newKeys = Object.keys(process.env).filter((key) => existingKeys.indexOf(key) === -1);
```
Comment (Nomad-AE) [inline distribution/src/autotask.spec.js:11]:
```suggestion
const newKeys = Object.keys(process.env).filter((key) => !existingKeys.includes(key));
```

File: distribution/src/autotask.js [hunk 39]
```diff
@@ -0,0 +1,220 @@
...
+
+  // // construct the Etherscan transaction link
+  const etherscanLink = `[TX](<https://etherscan.io/tx/${transactionHash}>)`;
+
+  // handle case for infinity % (no comp accrued before)
+  if (compAccrued.isEqualTo(0) || compDistributed.isEqualTo(0)) {
+    return `${etherscanLink} 🌊 **${receiverFormatted}** previously had no **COMP** accrued and was just distributed **COMP** tokens`;
+  }
+
+  // % disctribution % is calculated by compDistributed / compAccrued * 100
```
Comment (Nomad-AE) [inline distribution/src/autotask.js:28]:
```suggestion
  // % distribution % is calculated by compDistributed / compAccrued * 100
```

File: distribution/src/autotask.spec.js [hunk 40]
```diff
@@ -0,0 +1,175 @@
...
+});
+
+const autotaskConfig = require('../autotask-config.json');
+
+const { jsonRpcUrl } = autotaskConfig;
+
+// create a provider that will be injected as the Defender Relayer provider
+const mockProvider = new ethers.providers.JsonRpcBatchProvider(jsonRpcUrl);
+jest.mock('defender-relay-client/lib/ethers', () => ({
+  DefenderRelayProvider: jest.fn().mockReturnValue(mockProvider),
```
Comment (Nomad-AE) [inline distribution/src/autotask.spec.js:24]: Add  `defender-relay-client` to the `package.json` > `dependencies` list

File: governance/src/autotask.js [hunk 41]
```diff
@@ -0,0 +1,311 @@
+// eslint-disable-next-line import/no-extraneous-dependencies
```
Comment (Nomad-AE) [inline governance/src/autotask.js:1]:
```suggestion
```
Not necessary.

File: governance/src/autotask.js [hunk 42]
```diff
@@ -0,0 +1,311 @@
...
+      }
+
+      if (reason !== '') {
+        message += `\n\`\`\`${reason}\`\`\``;
+      }
+      break;
+    case 'AE-COMP-GOVERNANCE-PROPOSAL-CANCELED':
+      ({ id } = metadata);
+      proposalName = await getProposalTitle(id);
+      message = `**Canceled Proposa**l ${proposalName} ${noEntryEmoji}`;
```
Comment (Nomad-AE) [inline governance/src/autotask.js:219]:
```suggestion
      message = `**Canceled Proposal** ${proposalName} ${noEntryEmoji}`;
```

File: governance/src/autotask.spec.js [hunk 43]
```diff
@@ -0,0 +1,173 @@
+const ethers = require('ethers');
```
Comment (Nomad-AE) [inline governance/src/autotask.spec.js:1]: Add  `ethers` to the `package.json` > `dependencies` list

File: governance/src/autotask.js [hunk 44]
```diff
@@ -0,0 +1,311 @@
...
+  console.log('autotaskEvent');
+  console.log(JSON.stringify(autotaskEvent, null, 2));
+
+  const { secrets } = autotaskEvent;
+  if (secrets === undefined) {
+    return {};
+  }
+
+  // ensure that there is a DiscordUrl secret
+  const { TestingDiscordUrl: discordUrl } = secrets;
```
Comment (Nomad-AE) [inline governance/src/autotask.js:255]:
```suggestion
  const { FortaSentinelTestingDiscord: discordUrl } = secrets;
```

File: governance/src/autotask.js [hunk 45]
```diff
@@ -0,0 +1,311 @@
...
+      transactionHash,
+      agent: {
+        id: agentId,
+      },
+    },
+  } = alert;
+
+  // retrieve the metadata from the Forta public API
+  let alerts = await getFortaAlerts(agentId, transactionHash);
+  alerts = alerts.filter((alertObject) => alertObject.hash === hash);
```
Comment (Nomad-AE) [inline governance/src/autotask.js:291]:
```suggestion
```
I'm not seeing `hash` inside of the `alertObject` so this line filters out all results, making `alerts` empty. Producing zero discord messages. Commenting out or removing this line, allows it to work as intended.

Comment (Nomad-AE) [reply governance/src/autotask.js:291]: Block 14769156 for testing

File: governance/src/autotask.spec.js [hunk 46]
```diff
@@ -0,0 +1,173 @@
...
+const axios = require('axios');
+
+// grab the existing keys before loading new content from the .env file
+const existingKeys = Object.keys(process.env);
+require('dotenv').config();
+
+// now filter out all of the existing keys from what is currently in the process.env Object
+const newKeys = Object.keys(process.env).filter((key) => existingKeys.indexOf(key) === -1);
```
Comment (Nomad-AE) [inline governance/src/autotask.spec.js:9]:
```suggestion
const newKeys = Object.keys(process.env).filter((key) => !existingKeys.includes(key));
```
Nit

File: distribution/src/autotask.spec.js [hunk 47]
```diff
@@ -0,0 +1,175 @@
+// DEFENDER FORTA SENTINEL AUTOTASK TESTING //
```
Comment (Nomad-AE) [inline distribution/src/autotask.spec.js:1]: Add `.gitignore` file for `autotask-config.json`

File: asset-monitor/src/autotask.spec.js [hunk 48]
```diff
@@ -0,0 +1,153 @@
+// DEFENDER FORTA SENTINEL AUTOTASK TESTING //
```
Comment (Nomad-AE) [inline asset-monitor/src/autotask.spec.js:1]: Add `.gitignore` file for `autotask-config.json`

File: asset-monitor/src/autotask.spec.js [hunk 49]
```diff
@@ -0,0 +1,153 @@
...
+const ethers = require('ethers');
+
+// grab the existing keys before loading new content from the .env file
+const existingKeys = Object.keys(process.env);
+require('dotenv').config();
```
Comment (Nomad-AE) [inline asset-monitor/src/autotask.spec.js:6]: Add  `defender-relay-client` to the `package.json` > `dependencies` list

File: asset-monitor/src/autotask.js (hunk 1 above)
Comment (Nomad-AE) [reply asset-monitor/src/autotask.js:5]: Also needle is missing from  the `package.json` > `dependencies` list

File: asset-monitor/src/autotask.js [hunk 50]
```diff
@@ -0,0 +1,186 @@
...
+
+  return `${etherscanLink} 🆙 Underlying asset for the **${cTokenSymbol}** cToken contract was upgraded`;
+}
+
+// post to discord
+async function postToDiscord(url, message) {
+  let response;
+  try {
+    // perform the POST request
+    response = needle.post(url, { content: message }, { json: true });
```
Comment (Nomad-AE) [inline asset-monitor/src/autotask.js:21]: Replace `needle` with `axios`

File: asset-monitor/README.md [hunk 51]
```diff
@@ -25,4 +25,38 @@ a cToken's underlying asset may have changed.
...
+- Create an `autotask-config.json` file with the following format:
+```
+{
+  "jsonRpcUrl": "https://your.preferred.json.rpc.endpoint/with/api-key",
+  "agentId": "0xFORTAAGENTIDHERE",
+  "startBlockNumber": <integer_for_starting_block_number>,
+  "endBlockNumber": <integer_for_ending_block_number>
+}
+```
+- Create a .env file that contains the name of your discord webook and the url for it:
```
Comment (Nomad-AE) [inline asset-monitor/README.md:61]:
```suggestion
- Create a .env file that contains the name of your discord webhook and the URL for it:
```

**Review by arbitrary-nomad — COMMENTED** [review]
A few files need to be linted and I found a few typos. Some of the files use needle and others use axios. I recommend axios for everything for consistency across all of our repos. There were also a few .gitignores and packages missing. All seems like pretty easy fixes. Other than that, the discord bot fires off as expected.

File: ctoken-monitor/src/autotask.js [hunk 52]
```diff
@@ -0,0 +1,330 @@
...
+    const { metadata, description } = alertData;
+    const { eventName, contractAddress, cTokenSymbol } = metadata;
+    let decimals;
+    let symbol;
+    if (cTokenSymbol === 'cETH') {
+      decimals = 18;
+      symbol = 'ETH';
+    } else {
+      ({ decimals, symbol } = await getDecimalsAndSymbol(contractAddress, provider));
+    }
```
[the remark this answers is not in the document]
Comment (darian-ae) [reply ctoken-monitor/src/autotask.js:312]: moved!

File: governance/src/autotask.js (hunk 45 above)
Comment (darian-ae) [reply governance/src/autotask.js:291]: thanks changed!

File: ctoken-monitor/src/autotask.spec.js (hunk 15 above)
Comment (darian-ae) [reply ctoken-monitor/src/autotask.spec.js:162]: I looked into this a bit more as well and didn't find a proper solution. I can make a note or a small ticket about it, but it looks ok to me if it's firing alerts to discord

Comment (darian-ae) [reply ctoken-monitor/src/autotask.spec.js:162]: looked into it some more and it works with a smaller range of blocks:

   "startBlockNumber": 14788000,
    "endBlockNumber": 14790000

Comment (Nomad-AE) [reply ctoken-monitor/src/autotask.spec.js:162]: I think I set it to 1 block range and it was still having issues. I'll check it again in a bit. If the smaller block range works, then cool.

Comment (Nomad-AE) [reply ctoken-monitor/src/autotask.spec.js:162]: Yep that works!

**Review by arbitrary-nomad — APPROVED** [review]
Looks good.

File: asset-monitor/package.json [hunk 53]
```diff
@@ -22,7 +22,9 @@
     "bignumber.js": "^9.0.2",
     "forta-agent": "^0.0.39",
     "web3": "^1.7.1",
-    "web3-eth": "^1.7.1"
+    "web3-eth": "^1.7.1",
+    "defender-relay-client": "^1.19.0",
+    "dotenv": "^16.0.0"
```
Comment (arbitrary-sparrow) [inline asset-monitor/package.json:27]: looks like the `needle` dependency is missing

Comment (arbitrary-sparrow) [inline asset-monitor/package.json:27]: looks like the `needle` dependency is missing

**Comment by arbitrary-sparrow** [comment]
For the Autotasks that use `needle` there needs to be extra logic to 'roll-up' that dependency into the code that gets run in Defender. This is because Defender pre-installs only a handful of packages into an AWS lambda instance for an Autotask to use: https://docs.openzeppelin.com/defender/autotasks#environment

**Comment by zionjr0** [comment]
Cool

**Pull request merged**